Add webhook/ with webhook JSON config and deploy.sh with docs

humphd opened the pull request:
Closes #77
Closes #76
Closes #74
Closes #50

This finishes the work begun in #57 to get our deployment webhook in place.  ITS wants it to run outside of Docker, and has already installed `webhook` on the manager nodes.

This is the first of at least 2 parts to get this completed.  I need to land the major bits here, then test it and adjust and document things in a follow-up.

The main points here are as follows:

1. On staging (and later production), we'll run a [webhook server](https://github.com/adnanh/webhook).  It will listen on :8000 and ITS will route anything that comes to `/hooks/*` to this server.
2. The webhook will listen for `deploy` messages from GitHub, which will include 2 things.  First a new `tag` to use with Docker.  When we update to a new version on `main`, this will send that image tag so it can be deployed; second, we'll send a digest created with a secret.  Our deployment servers will only respond to webhooks that can sign the body (i.e., no one else can hit this endpoint)
3. Assuming the webhook can be verified by the server (calculating the secret + body to see if it matches the digest), the `deploy.sh` script will be run.  The new `tag` will be sent along as an ENV var.  Docker will try to update the starchart image to the new tag.

As we discussed in #291, I'm not doing any database migration/syncing in here.  Any changes to the database schema would require a manual shut-down and update of the DB first.  It would be good to get to an automated version of this some day, but not in this PR.

humphd commented on .github/workflows/ci.yaml:
```diff
@@ -354,5 +349,16 @@ jobs:
+        with:
+          url: https://mycustomdomain-dev.senecacollege.ca/hooks/deploy
+          hmacSecret: ${{ env.HMAC_SECRET }}
+          data: '{"tag": "${{ env.IMAGE }}:sha-${{ env.GITHUB_SHA_SHORT }}"}'
```
Don't send image here, only the new tag.

Eakam1007 replied:
Apart from this, and one small question looks good

sfrunza13 commented on .github/workflows/ci.yaml:
```diff
@@ -354,5 +349,16 @@ jobs:
+            ${{ env.IMAGE }}:latest
+
+      - name: Deploy to Staging via Webhook
+        uses: navied/secure-actions-webhook@0.2.0
```
If I am correct this was written by senior staff engineer @ shopify, last updated a few years ago, one star on the repo. How did you find this? 👀

humphd replied:
https://github.com/marketplace?type=actions. I needed an HMAC SHA256 webhook action, and there are precious few for some reason.

sfrunza13 commented on webhook/deploy.sh:
```diff
@@ -0,0 +1,22 @@
+SERVICE=starchart_mycustomdomain
+
+# We expect to receive the new image tag via the TAG environment variable
+if [[ -z "$TAG" ]]; then
```
Idk how to read this, what is -z?

Eakam1007 replied:
Seems like this is to check it $TAG is null

sfrunza13 commented on .github/workflows/ci.yaml:
```diff
@@ -314,32 +314,27 @@ jobs:
         uses: actions/checkout@v3
 
+      - name: Define GITHUB_SHORT_SHA
+        run: echo "GITHUB_SHA_SHORT=$(echo $GITHUB_SHA | cut -c 1-6)" >> $GITHUB_ENV
```
I don't think this really makes a difference but aren't the characters on the tail end of a sha the most meaningful?

humphd replied:
You can take any hash and truncate it to use only the first part of it as a short-hand.  This is what GitHub does when they show my commit as `93a5088` vs. `93a508896294bdd9f74b1d4ca92a57c49999f34e`

sfrunza13 reviewed (approved):
I don't think I am going to be able to find a fault with this. After everything in the needs occurs add IMAGE to env, then build and push an image with new tags, use navied/secure-actions-webhook@0.2.0 to post to hooks safely, then deploy.sh logic which should use the tag passed to it through data as you already mentioned yourself so that will be changed for the better. I think it's good.

Eakam1007 commented on webhook/hooks.json:
```diff
@@ -0,0 +1,24 @@
+    "trigger-rule": {
+      "match": {
+        "type": "payload-hmac-sha256",
+        "secret": "TODO: update me to match secret stored in GitHub Actions...",
```
Wanted to confirm, Is this secret supposed to be a path to a secret file or something like that? Or is this going to be manually updated later (and not comnitted to git)? Doesn't seems like a secret if we commit this here

sfrunza13 replied:
good point

Eakam1007 replied:
So, will you ssh into the server to change this manually?

Eakam1007 reviewed (dismissed):
Just a small question out of curiosity, but looks good

humphd commented:
> Just a small question out of curiosity, but looks good

Correct, these files won't be used as is. They are a template for manually setting this up on staging.

humphd commented:
I've updated this again.  I've added the `systemd` startup unit file, so the webhook runs as a service (it's already installed on staging).  I think this is ready to test!

humphd commented:
Thanks for the review!  I'm going to try this now...

humphd commented:
Crap, I got my order wrong, I have to do another PR to make this happen.
